libass/ass.c

```diff
@@ -769,6 +770,7 @@ static int process_events_line(ASS_Track *track, char *str)
         skip_spaces(&p);
         free(track->event_format);
         track->event_format = strdup(p);
+        if (!track->event_format) return -1;
```

Comment:
To be honest, shouldn’t we reject the whole track if we have failed to allocate memory for a `Format:` line? (Barring a particularly weird file that has multiple `Format:` lines where we succeed in reading the last one… which should in practice only happen when libass is being fuzzed.) It is supposed to affect the very parsing process, so continuing with an arbitrary fallback format line seems wrong.

Also note that `custom_format_line_compatibility` never receives `NULL` in the existing code: it is given `p`, not `strdup(p)`.

Reply:
> eject the whole track if we have failed to allocate memory for a `Format:` line?

I'd love to do that, but right now this seems impossible for API-users using `ass_process_codec_private` and `ass_process_chunk`. With a proper error api in the future, this could be changed.

What could possibly be done now is to let track creation fail for all API-calls that allow it, if **any** allocation fails during script parsing.

Reply:
We _could_ just discard/ignore/skip lines that we don’t have a format for. It might be better than parsing them incorrectly. But we should make a distinction between “there was no Format line to begin with and we’re falling back to a default” and “there was one, but we failed to allocate memory for it”.

Of course, if/when we finally stop honouring custom format lines except FFmpeg’s, this will largely become moot. Of course, this “should” largely be moot to begin with as the strdup shouldn’t fail in practice…

If you don’t think this is worthwhile/good right now, it’s fine to leave it as is.

Reply:
My preferred way would be to leave at is for now, but change it to mark the whole track as garbage on format-line failure when the proper error API is introduced soon™.


> Of course, if/when we finally stop honouring custom format lines except FFmpeg’s

For the record, FFmpeg had multiple custom format lines and custom format lines are also used for SSA-compatibility. Continuing to accept all format lines with Text at the end is probably cleaner and easier as far as the code is concerned than trying to special case SSA's, `ASS (v4+)`'s, `ASS (v4++)`'s and FFmpeg's format lines.

Reply:
I was thinking of using the FFmpeg-authoring detection code you wrote for ScaledBorderAndShadow.

I recall complaints about files with incorrect Format lines that work in VSFilter (although I can’t immediately find an issue for it on our Google Code and GitHub trackers). Honouring Format lines that end in Text seems like it would defeat the point. But either way, that’s out of scope for this PR.

Let’s leave it as is for now, then.